# Recurrent Neural Networks

Gabs DiLiegro, London Kasper, Carys LeKander

Dataset: https://www.kaggle.com/datasets/uciml/sms-spam-collection-dataset?select=spam.csv

## Preparation

Our dataset is a collection of text messages that are identified as normal or spam texts.

In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv('spam.csv',encoding='latin-1')
df = df.drop(['Unnamed: 2','Unnamed: 3', 'Unnamed: 4'], axis=1)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   v1      5572 non-null   object
 1   v2      5572 non-null   object
dtypes: object(2)
memory usage: 87.2+ KB


The first column of our dataset identifies if it is a spam text. We used label encoding below to change the column to be 0 for a normal text and 1 for a spam text. 

In [2]:
from sklearn import preprocessing
from sklearn.preprocessing import LabelEncoder

le = preprocessing.LabelEncoder()
le.fit(df['v1'])
df['v1'] = le.transform(df['v1'])

df.head()

v1                                                 v2
0   0  Go until jurong point, crazy.. Available only ...
1   0                      Ok lar... Joking wif u oni...
2   1  Free entry in 2 a wkly comp to win FA Cup fina...
3   0  U dun say so early hor... U c already then say...
4   0  Nah I don't think he goes to usf, he lives aro...

Looking at the totals for each category, we can see that there are many more normal text messages than spam messages in our dataset. 

In [3]:
totals = df['v1'].value_counts()
totals

0    4825
1     747
Name: v1, dtype: int64

The longest text message in our dataset is 171 which we will use for our max length during tokenization.


In [4]:
max_words = 0
for i in df.v2:
    if len(i.split()) > max_words:
        max_words = len(i.split())
max_words

171

For our tokenization, we converted each word to an integer using the Keras Tokenizer.

In [5]:
%%time
from tensorflow import keras
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences

NUM_TOP_WORDS = None # use entire vocabulary!
MAX_ART_LEN = max_words # maximum and minimum number of words

#tokenize the text
tokenizer = Tokenizer(num_words=NUM_TOP_WORDS)
tokenizer.fit_on_texts(df.v2)
# save as sequences with integers replacing words
sequences = tokenizer.texts_to_sequences(df.v2)

word_index = tokenizer.word_index
NUM_TOP_WORDS = len(word_index) if NUM_TOP_WORDS==None else NUM_TOP_WORDS
top_words = min((len(word_index),NUM_TOP_WORDS))
print('Found %s unique tokens. Distilled to %d top words.' % (len(word_index),top_words))

X = pad_sequences(sequences, maxlen=MAX_ART_LEN)
print('Shape of data tensor:', X.shape)
y = df.v1.astype(np.int32)

2022-12-14 03:12:49.690524: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


Found 8920 unique tokens. Distilled to 8920 top words.
Shape of data tensor: (5572, 171)
CPU times: user 4.75 s, sys: 698 ms, total: 5.45 s
Wall time: 6.78 s


For our business case, we want to notify users when they recieve a text that may be spam. If we misclassify something as spam, the user may be slightly annoyed. However, if we do not clasify something as spam when it is, it could lead to the user trusting a fraudulent text. Therefore we want to minimize the number of false negative (number of spam texts classified as normal texts), so we will use recall to measure our algorithm's performance.

We chose to use a Stratified 10-fold cross validation for dividing our data into training and testing. We chose this method because our dataset is smaller (5572) and there are not many examples of spam texts (747). We want to make sure that we represent the entire dataset in equal proportion.

MAYBE ADD MORE EXPLANATION TO THIS

In [6]:
from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=1)
y_train = y_train.values.astype(np.int32)
y_test = y_test.values.astype(np.int32)
kfold = StratifiedKFold(n_splits=10).split(X_train, y_train)

## Modeling

***Investigate at least two different recurrent network architectures (perhaps LSTM and GRU). Alternatively, you may also choose one recurrent network and one convolutional network. Be sure to use an embedding layer (try to use a pre-trained embedding, if possible). Adjust hyper-parameters of the networks as needed to improve generalization performance (train a total of at least four models). Discuss the performance of each network and compare them.***

In [7]:
%%time
EMBED_SIZE = 100
# the embed size should match the file you load glove from
embeddings_index = {}
f = open('large_data/glove.6b/glove.6B.100d.txt')
# save key/array pairs of the embeddings
#  the key of the dictionary is the word, the array is the embedding
for line in f:
    values = line.split()
    word = values[0]
    coefs = np.asarray(values[1:], dtype='float32')
    embeddings_index[word] = coefs
f.close()

print('Found %s word vectors.' % len(embeddings_index))

# now fill in the matrix, using the ordering from the
#  keras word tokenizer from before
found_words = 0
embedding_matrix = np.zeros((len(word_index) + 1, EMBED_SIZE))
for word, i in word_index.items():
    embedding_vector = embeddings_index.get(word)
    if embedding_vector is not None:
        # words not found in embedding index will be ALL-ZEROS
        embedding_matrix[i] = embedding_vector
        found_words = found_words+1

print("Embedding Shape:",embedding_matrix.shape, "\n",
      "Total words found:",found_words, "\n",
      "Percentage:",100*found_words/embedding_matrix.shape[0])

Found 400000 word vectors.
Embedding Shape: (8921, 100) 
 Total words found: 6518 
 Percentage: 73.06355789709674
CPU times: user 10.1 s, sys: 305 ms, total: 10.4 s
Wall time: 10.5 s


In [8]:
from tensorflow.keras.models import Sequential, Model
from tensorflow.keras.layers import Dense, Input, LSTM, GRU
from tensorflow.keras.layers import Embedding

In [9]:
# save this embedding now
embedding_layer = Embedding(len(word_index) + 1,
                            EMBED_SIZE,
                            weights=[embedding_matrix],# here is the embedding getting saved
                            input_length=MAX_ART_LEN,
                            trainable=False)

In [10]:
lstm1 = Sequential()
lstm1.add(embedding_layer)
lstm1.add(LSTM(100,dropout=0.2, recurrent_dropout=0.2))
lstm1.add(Dense(1, activation='sigmoid'))

print(lstm1.summary())

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, 171, 100)          892100    
                                                                 
 lstm (LSTM)                 (None, 100)               80400     
                                                                 
 dense (Dense)               (None, 1)                 101       
                                                                 
Total params: 972,601
Trainable params: 80,501


2022-12-14 03:13:07.073628: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


Non-trainable params: 892,100
_________________________________________________________________
None


In [11]:
lstm2 = Sequential()
lstm2.add(embedding_layer)
lstm2.add(LSTM(100,dropout=0.2, recurrent_dropout=0.3))
lstm2.add(Dense(1, activation='sigmoid'))

print(lstm2.summary())

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, 171, 100)          892100    
                                                                 
 lstm_1 (LSTM)               (None, 100)               80400     
                                                                 
 dense_1 (Dense)             (None, 1)                 101       
                                                                 
Total params: 972,601
Trainable params: 80,501
Non-trainable params: 892,100
_________________________________________________________________
None


In [12]:
gru1 = Sequential()
gru1.add(embedding_layer)
gru1.add(GRU(100,dropout=0.2, recurrent_dropout=0.2))
gru1.add(Dense(1, activation='sigmoid'))

print(gru1.summary())

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, 171, 100)          892100    
                                                                 
 gru (GRU)                   (None, 100)               60600     
                                                                 
 dense_2 (Dense)             (None, 1)                 101       
                                                                 
Total params: 952,801
Trainable params: 60,701
Non-trainable params: 892,100
_________________________________________________________________
None


In [13]:
gru2 = Sequential()
gru2.add(embedding_layer)
gru2.add(GRU(100,dropout=0.2, recurrent_dropout=0.3))
gru2.add(Dense(1, activation='sigmoid'))

print(gru2.summary())

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, 171, 100)          892100    
                                                                 
 gru_1 (GRU)                 (None, 100)               60600     
                                                                 
 dense_3 (Dense)             (None, 1)                 101       
                                                                 
Total params: 952,801
Trainable params: 60,701
Non-trainable params: 892,100
_________________________________________________________________
None


**Use the method of train/test splitting and evaluation criteria that you argued for at the beginning of the lab. Visualize the results of all the RNNs you trained.  Use proper statistical comparison techniques to determine which method(s) is (are) superior.**

In [14]:
from sklearn.metrics import roc_curve, auc

mean_fpr = np.linspace(0, 1, 100)

def roc(model, test, train, mean_tpr):
    probas = model.predict(X_train[test], verbose=0)
    fpr, tpr, thresholds = roc_curve(y_train[test], probas, pos_label=1)
    mean_tpr += np.interp(mean_fpr, fpr, tpr)
    mean_tpr[0] = 0.0
    return mean_tpr

In [15]:
%%time
acc_per_fold1 = []
recall_per_fold1 = []
loss_per_fold1 = []
mean_tpr1 = 0.0

acc_per_fold2 = []
recall_per_fold2 = []
loss_per_fold2 = []
mean_tpr2 = 0.0

acc_per_fold3 = []
recall_per_fold3 = []
loss_per_fold3 = []
mean_tpr3 = 0.0

acc_per_fold4 = []
recall_per_fold4 = []
loss_per_fold4 = []
mean_tpr4 = 0.0

# K-fold Cross Validation model evaluation
for k, (train, test) in enumerate(kfold):

    # Compile the model
    lstm1.compile(loss='binary_crossentropy',
            optimizer='rmsprop',
            metrics=['accuracy', keras.metrics.Recall()])
    lstm2.compile(loss='binary_crossentropy',
            optimizer='rmsprop',
            metrics=['accuracy', keras.metrics.Recall()])
    gru1.compile(loss='binary_crossentropy',
            optimizer='rmsprop',
            metrics=['accuracy', keras.metrics.Recall()])
    gru2.compile(loss='binary_crossentropy',
            optimizer='rmsprop',
            metrics=['accuracy', keras.metrics.Recall()])


    # Generate a print
    print(f'Training for fold {k+1} ...')

    # Fit data to model
    history1 = lstm1.fit(X_train[train], y_train[train], epochs=3, batch_size=64, validation_data=(X_train[test], y_train[test],), verbose=0)
    history2 = lstm2.fit(X_train[train], y_train[train], epochs=3, batch_size=64, validation_data=(X_train[test], y_train[test],), verbose=0)
    history3 = gru1.fit(X_train[train], y_train[train], epochs=3, batch_size=64, validation_data=(X_train[test], y_train[test],), verbose=0)
    history4 = gru2.fit(X_train[train], y_train[train], epochs=3, batch_size=64, validation_data=(X_train[test], y_train[test],), verbose=0)

    mean_tpr1 = roc(lstm1, test, train, mean_tpr1)
    mean_tpr2 = roc(lstm2, test, train, mean_tpr2)
    mean_tpr3 = roc(gru1, test, train, mean_tpr3)
    mean_tpr4 = roc(gru2, test, train, mean_tpr4)

    # Generate generalization metrics
    scores1 = lstm1.evaluate(X_train[test], y_train[test], verbose=0)
    scores2 = lstm2.evaluate(X_train[test], y_train[test], verbose=0)
    scores3 = gru1.evaluate(X_train[test], y_train[test], verbose=0)
    scores4 = gru2.evaluate(X_train[test], y_train[test], verbose=0)

    print(f'Score for LSTM1: {lstm1.metrics_names[0]} of {scores1[0]}; {lstm1.metrics_names[1]} of {scores1[1]*100}%; {lstm1.metrics_names[2]} of {scores1[2]*100}%')
    print(f'Score for LSTM2: {lstm2.metrics_names[0]} of {scores2[0]}; {lstm2.metrics_names[1]} of {scores2[1]*100}%; {lstm2.metrics_names[2]} of {scores2[2]*100}%')
    print(f'Score for GRU1: {gru1.metrics_names[0]} of {scores3[0]}; {gru1.metrics_names[1]} of {scores3[1]*100}%; {gru1.metrics_names[2]} of {scores3[2]*100}%')
    print(f'Score for GRU2: {gru2.metrics_names[0]} of {scores4[0]}; {gru2.metrics_names[1]} of {scores4[1]*100}%; {gru2.metrics_names[2]} of {scores4[2]*100}%')

    loss_per_fold1.append(scores1[0])
    acc_per_fold1.append(scores1[1] * 100)
    recall_per_fold1.append(scores1[2]*100)
    loss_per_fold2.append(scores2[0])
    acc_per_fold2.append(scores2[1] * 100)
    recall_per_fold2.append(scores2[2]*100)
    loss_per_fold3.append(scores3[0])
    acc_per_fold3.append(scores3[1] * 100)
    recall_per_fold3.append(scores3[2]*100)
    loss_per_fold4.append(scores4[0])
    acc_per_fold4.append(scores4[1] * 100)
    recall_per_fold4.append(scores4[2]*100)


Training for fold 1 ...
Score for LSTM1: loss of 0.09335790574550629; accuracy of 95.96412777900696%; recall of 93.44262480735779%
Score for LSTM2: loss of 0.08601104468107224; accuracy of 97.98206090927124%; recall_1 of 86.8852436542511%
Score for GRU1: loss of 0.06690991669893265; accuracy of 98.20627570152283%; recall_2 of 91.80327653884888%
Score for GRU2: loss of 0.08441730588674545; accuracy of 97.98206090927124%; recall_3 of 86.8852436542511%
Training for fold 2 ...
Score for LSTM1: loss of 0.07476350665092468; accuracy of 97.98206090927124%; recall_4 of 86.8852436542511%
Score for LSTM2: loss of 0.10344936698675156; accuracy of 96.41255736351013%; recall_5 of 75.40983557701111%
Score for GRU1: loss of 0.10055604577064514; accuracy of 96.8609869480133%; recall_6 of 78.68852615356445%
Score for GRU2: loss of 0.07904259115457535; accuracy of 97.53363132476807%; recall_7 of 83.60655903816223%
Training for fold 3 ...
Score for LSTM1: loss of 0.021923579275608063; accuracy of 99.5515